### PR TITLE
[CR] Hide uncraftable recipes in crafting interface and hidden_if_uncraftable flag to recipes

### PR DESCRIPTION
**Pull request by tivec**
This allows the player to hide recipes in the crafting interface that are not possible to craft due to missing ingredients. This sohuld result in a less cluttered feel when the player has a lot of recipes.

It does not save the flag state when the crafting interface is closed.

The "hidden_if_uncraftable" flag, available for recipes, will make it so the recipe is always hidden until the player has all the components available to craft the item.

### CODE REVIEW

**Comment by kevingranade** [comment]
I never got a good explanation for why one would want to always hide recipe details when you don't have the ingredients, it seems like it would very much damage your ability to figure out how to make the item in question.

**Comment by Lorith** [comment]
This was actually something I had been asking about, more for modding rather crazy things in.  I have been working on and off at making a supply drone mod, and in the finished state it will have four sets of recipes for at least seventeen resources both sending and receiving.  That leads to a HUGE number of recipes being shown, most of which are useless at any point, and it could massively streamline the list by only showing the ones that are possible, such as only the outbound or only the inbound.  For things such as that, showing 136 recipes all at once when half or more are useless is a bit crazy.  Ideally, it would be nice to be able to set a recipe as impossible to memorize and automatically usable just from having an item without having to read it.  Then trigger items could be used in such cases.

**Comment by kevingranade** [comment]
So it's specifically for managing a problem that only arises if you're
using the crafting system for something it isn't meant to handle.  That's
not a compelling reason to have it in the code.
You *should* be able to mark the recipes as not memorizable, and access
them directly from an iuse instead of having the player learn them.

**Comment by BevapDin** [comment]
> The "hidden_if_uncraftable" flag, available for recipes, will make it so the recipe is always hidden until the player has all the components available to craft the item.

And how do I know what I still need and what I already have? Consider a crafting area stocked with stuff of all kind. I would have to look at the json to find out what I need, that I have to check each component manually for the required amount.

> This allows the player to hide recipes in the crafting interface that are not possible to craft due to missing ingredients. This sohuld result in a less cluttered feel when the player has a lot of recipes.

Recipes that have enough materials are printed in white (other are printed in gray), and are move to the top of the list. Simply ignore every recipe below the last white recipe, they are all unavailable.

**Comment by tivec** [comment]
I have another angle on this. Let's say that we make an iuse for this - one where the player activates a certain item that will be able to call in the drones. This iuse could piggyback on the crafting system to do all the hard work (finding the correct items for crafting etc). The recipes would be hidden in the crafting system, because it's not relevant for them to be there.

**Comment by BevapDin** [comment]
Than load a and store the recipes within the iuse - make a specific `iuse_actor` class that holds a list of the related recipes. You probably only need some `requirements` objects and not a full blown `recipe` object anyway. And even if you need a full recipe object with all its additional data, just load it into your own list of recipes, not into the global recipe list, you need to split the recipe loading function `load_recipe` into a "load" and a "add to global recipe list" function.

In short: why add the recipe to the global recipe list when you'll never use it from there.

**Comment by tivec** [comment]
@BevapDin The crafting system is perfectly fine at making the items, and riding piggyback on that is in my opinion smarter than duplicating code... But, I'll heed your recommendations and look into using iuse_actors and iuse. Is there a segment of code that uses the same style of splitting of recipes instead of loading them to the global list? My ability with C++ is really not good enough to make specific iuse_actor classes for this thing, so I'll probably have to leave it to someone else.

**Comment by KA101** [comment]
OK, closing this as unlikely to merge in present config.  Sorry, Tivec.
